src/software/ai/evaluation/q_learning/attacker_mdp_feature_extractor.cpp

```diff
@@ -0,0 +1,43 @@
+#include "software/ai/evaluation/q_learning/attacker_mdp_feature_extractor.h"
+
+#include <math.h>
+
+#include "software/ai/evaluation/possession.h"
+
+AttackerMdpFeatureExtractor::AttackerMdpFeatureExtractor()
+    : FeatureExtractor({ballXPositionFeature, bestPassRatingFeature, bestShotOpenAngleFeature})
+{
+}
+
+double AttackerMdpFeatureExtractor::ballXPositionFeature(const AttackerMdpState& state)
+{
+    const double x_pos = state.world_ptr->ball()
+        .estimateFutureState(Duration::fromSeconds(1.0)).position().x();
+    const double field_x_length   = state.world_ptr->field().totalXLength();
+    const double normalized_x_pos = std::max(x_pos, 0.0) / (field_x_length / 2);
+    return normalized_x_pos;
+}
+
+double AttackerMdpFeatureExtractor::bestPassRatingFeature(const AttackerMdpState& state)
+{
+    return state.strategy->getBestPass().rating;
+}
+
+double AttackerMdpFeatureExtractor::bestShotOpenAngleFeature(
+    const AttackerMdpState& state)
+{
+    std::optional<Robot> attacker = getRobotWithEffectiveBallPossession(
+        state.world_ptr->friendlyTeam(), state.world_ptr->ball(),
+        state.world_ptr->field());
+
+    if (attacker)
+    {
+        std::optional<Shot> best_shot = state.strategy->getBestShot(*attacker);
+        if (best_shot)
+        {
+            return best_shot->getOpenAngle().toRadians() / (M_PI / 4);
```

Comment:
am I understanding this correctly? Does this incentivize open shots with larger open angles?

Reply:
Yeah somewhat. If a feature value is large and its associated weight for a given action is large, then this feature will more heavily contribute to the Q-value for that action. So if we have a large open shot angle AND the shot angle feature / shoot action weight is large (which got large over time after the attacker learned this feature was important for a successful shoot action), then we will likely choose the shoot action because the Q-value will be larger